DOC/man/ekb_insert.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH EKB_INSERT "1" "May 2024" "ekb_insert 3.1.0" "User Commands"
.SH NAME
ekb_insert \- manual page for ekb_insert 3.1.0
.SH SYNOPSIS
.B ekb_insert
[\fI\,options\/\fR] [\fI\,names\/\fR]
.SH DESCRIPTION
ekb_insert 3.1.0
.PP
Insert example files into an E knowledge base. Each non\-option argument
is considered as one individual example file. For most applications
this is obsolete, use ekb_ginsert instead.
.PP
Options
.HP
\fB\-h\fR
.HP
\fB\-\-help\fR
.IP
Print a short description of program usage and options.
.HP
\fB\-V\fR
.HP
\fB\-\-version\fR
.IP
Print the version number of the program.
.HP
\fB\-v\fR
.HP
\fB\-\-verbose[=\fR<arg>]
.IP
Verbose comments on the progress of the program. The short form or the
long form without the optional argument is equivalent to \fB\-\-verbose\fR=\fI\,1\/\fR.
.HP
\fB\-n\fR <arg>
.HP
\fB\-\-name=\fR<arg>
.IP
Give the name of the new problem. If not given, the program will take the
name of the first input file, or, if <stdin> is read, a name of the form
\&'__problem__i', where i is magically computed  from the existing
knowledge base.
.HP
\fB\-k\fR <arg>
.HP
\fB\-\-knowledge\-base=\fR<arg>
.IP
Select the knowledge base. If not given, select E_KNOWLEDGE.
.SH "REPORTING BUGS"
.PP
Report bugs to <schulz@eprover.org>. Please include the following, if
possible:
.PP
* The version of the package as reported by \fBeprover \-\-version\fR.
.PP
* The operating system and version.
.PP
* The exact command line that leads to the unexpected behaviour.
.PP
* A description of what you expected and what actually happened.
.PP
* If possible all input files necessary to reproduce the bug.
.SH COPYRIGHT
Copyright 1998\-2024 by Stephan Schulz, schulz@eprover.org,
and the E contributors (see DOC/CONTRIBUTORS).
.PP
This program is a part of the distribution of the equational theorem
prover E. You can find the latest version of the E distribution
as well as additional information at
http://www.eprover.org
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program (it should be contained in the top level
directory of the distribution in the file COPYING); if not, write to
the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
Boston, MA  02111\-1307 USA
.PP
The original copyright holder can be contacted via email or as
.PP
Stephan Schulz
DHBW Stuttgart
Fakultaet Technik
Informatik
Lerchenstrasse 1
70174 Stuttgart
Germany


DOC/NEWS
This history describes the major differences between versions of the E
equational theorem prover.

Version 3.1.0 -> 3.2.X Puttabong Moondrop

- Added support for $distinct
- Disabled -xAuto and cut out most of the relics of the old auto-mode
  stuff
- Added option to select one of the built-in strategies
- Took out old clausifier
- Updated several options
- Debugged some higher-order problems

Version 3.0.03 -> 3.1.0

- Added disequality decomposition inference
- Fixed various bugs (e.g. in local rewriting)
- Added support for additional statistics
- Enabled printout of all built-in search strategies


Version 3.0 ->3.0.03 Shangri-La

- Fixed bug in activating clause selection heuristics
- Fixed several bugs in the higher-order module, mostly related to the
  difference between shared and unshared terms. This will need further
  work to catch all corner cases.
- Fixed several minor memory leaks and simplified management of parsed
  strings with the clb_permastrings module.

Version 3.0 (pre-release) ->3.0 Shangri-La

- Fixed various bugs in the HO preprocessor
- Fixed various bugs in the HO reasoner
- Improved the multicore scheduler

Version 2.6 ->3.0 Shangri-La (pe-release)

- New option --fw-subsumption-aggressive.
- Fixed bug in negative unit subsumption.
- Added new DAG-based evaluation functions
- Fixed status for incomplete literal weight functions
- Added support for UTF-8 in Strings and single-quoted symbols.
- Added multicore scheduler and schedule generation
- Replaced auto-mode code by new symbolic auto-mode
- Various minor improvements
- Added support for full (monomorphic) higher-order reasoning (PV)


Version 2.5 -> 2.6 Floral Guranse

- Much improved algorithm for finding maximal literals in long clauses
- Much improved algorithm for identifying dual (conflicting) literals
  in long clauses
- Added max literal limit for full equational tautology check (we
  should now really replace that with a congruence closure based
  approach!)
- Added DAGWeight() evaluation function (symbol-counts shared terms
  only once, with different options for sharing regions (terms,
  literals, clauses)
- Improved support for THF: E can now parse most of TH0 syntax
- (notable exceptions are symbols !!, ?? and @+), including lambdas.
  Note that on the reasoning side, E still does not support any
  advanced lambda reasoning, as lambdas are simply lifted and after
  parsing E is not aware of them.
- Improved support for TFX: E now supports $let and $ite expressions.



Version 2.4 -> 2.5 Avongrove

- Clausal problems are now classified before equational definitions
  expansion, which moves equational definition expansion under control
  auf the automatic modes)
- Strong rewrite (optionally instantiating unbound variables of the
  right hand side of equations considered for rewriting) has neem
  fixed and updeted to the sorted case.
- An improvement in strategy scheduling will now aovid repetition of
  the same strategy in cases where the data does not change from one
  level to the next, allowing for meaningful deeper schedules.
- New automatic modes
- Added variants to symbol precedence schemes (prefer/defer
  skolems/defined symbols/axiom-symbols/conjecture symbols).


Version 2.3 -> 2.4 Sandakphu

- Added random clause "evaluation"
- Added Diversity-based clause evaluation
- Added (simplified) transfinite orderings for literal comparision
- New automatic modes
- Support for CASC-27 LTB (with various alternative formalisations per
  problem)
- Slightly finer classificarion of problems (for improved UEQ
  performance)

Version 2.3pre -> 2.3 Gielle

- Made clausification more robust for extreme examples

Version 2.2 -> 2.3pre Gielle

- Merged support for lambda-free higher-order logic.
- Improved fingerprint indexing by excluding one more subcase for
  finding instances

Version 2.1 -> 2.2 Thurbo Moonlight

- Prover now always builds internal proof object (--proof-object only
  controls output)
- As a result, E can now detect ContradictoryAxioms if the proof state
  is unsatisfiable even without the conjecture.
- Moved from eager to lazy orphan deletion. If you don't know what
  this means: It's much cooler and simpler.
- Fixed and simplified processing of unprocessed clauses
- Refactored PicoSAT integration - PicoSAT is now linked into E
  proper, and used via its API. We are back to a single binary
  executable with no dependencies.
- Prover now enforces that fof, tff and tcf formulas are closed (no
  free variables allowed)
- Fixed condensation (somewhat).
- New automatic modes (but with very few current test results)


Version 2.0 -> 2.1 Maharani Hills

- Added type output for CNFed TFFs
- Added tcf sublanguage for clause-form TFFs
- Combined consecutive applications of the same quantifier into a
  single quantifier
- Improved stability and performance on TFF problems
- Changed internal workflow of unprocessed clauses so that they can be
  batch-evaluated (e.g. by a Deep Neural Network).
- Added PicoSAT and the ability to use PicoSAT as a SAT terminator
- Added support for flexible-arity derivation operations
- Removed obsolete eproof-scripts. Use --proof-object to generate a
  checkable proof.

Version 1.9.1 -> 2.0 Turzum

- Support for TFF (typed first-order form), thanks to Simon Cruanes
- Improved clausification (faster, with graceful degradation if
  miniscoping becomes too expensive)
- Improved auto-modes
- Improved Set-of-Support simulation
- Parsing and pruning preprocessing FOF now, even for CNF input
- Input clauses/formulas keep their original name (if any)
- Added options  --generated-limit and --tb-insert-limit to more
  closely limit effort by the prover in a deterministic way.

Version 1.9 -> 1.9.1 Sungma

- Improved proof object output
- Added optional recording of selected given clauses in the proof
  object
- New auto modes (including one bred heuristic)
- Fixed bug in formula simplification involving P=><=>$false

PROVER/ekb_insert.c
/*-----------------------------------------------------------------------

  File  : ekb_insert.c

  Author: Stephan Schulz

  Contents

  Insert an new training example file into a knowledge base.

  Copyright 1998, 1999-2006 by the author.
  This code is released under the GNU General Public Licence and
  the GNU Lesser General Public License.
  See the file COPYING in the main E directory for details..
  Run "eprover -h" for contact information.

  Created: Jul 28 16:21:33 MET DST 1999

  -----------------------------------------------------------------------*/

#include <cio_commandline.h>
#include <cio_output.h>
#include <cio_fileops.h>
#include <cle_kbinsert.h>
#include <e_version.h>

/*---------------------------------------------------------------------*/
/*                  Data types                                         */
/*---------------------------------------------------------------------*/

#define NAME    "ekb_insert"

typedef enum
{
   OPT_NOOPT=0,
   OPT_HELP,
   OPT_VERSION,
   OPT_VERBOSE,
   OPT_KB,
   OPT_NAME
}OptionCodes;



/*---------------------------------------------------------------------*/
/*                        Global Variables                             */
/*---------------------------------------------------------------------*/

OptCell opts[] =
{
   {OPT_HELP,
    'h', "help",
    NoArg, NULL,
    "Print a short description of program usage and options."},

   {OPT_VERSION,
    'V', "version",
    NoArg, NULL,
    "Print the version number of the program."},

   {OPT_VERBOSE,
    'v', "verbose",
    OptArg, "1",
    "Verbose comments on the progress of the program."},

   {OPT_NAME,
    'n', "name",
    ReqArg, NULL,
    "Give the name of the new problem. If not given, the program will "
    "take the name of the first input file, or, if <stdin> is read, a"
    " name of the form '__problem__i', where i is magically computed "
    " from the existing knowledge base."},

   {OPT_KB,
    'k',"knowledge-base",
    ReqArg, NULL,
    "Select the knowledge base. If not given, select E_KNOWLEDGE."},

   {OPT_NOOPT,
    '\0', NULL,
    NoArg, NULL,
    NULL}
};

char* ex_name = NULL;
char* kb_name = "E_KNOWLEDGE";
bool app_encode = false;

/*---------------------------------------------------------------------*/
/*                      Forward Declarations                           */
/*---------------------------------------------------------------------*/

CLState_p process_options(int argc, char* argv[]);
void print_help(FILE* out);

/*---------------------------------------------------------------------*/
/*                         Internal Functions                          */
/*---------------------------------------------------------------------*/


int main(int argc, char* argv[])
{
   CLState_p       state;
   DStr_p          name, store_file;
   FILE            *out;
   ExampleSet_p    proof_examples;
   AnnoSet_p       clause_examples;
   TB_p            annoterms;
   TypeBank_p      typebank;
   Sig_p           reserved_symbols;
   Scanner_p       in;
   char            defaultname[40];
   int             i;

   assert(argv[0]);
#ifdef STACK_SIZE
   INCREASE_STACK_SIZE;
#endif
   InitIO(NAME);

   state = process_options(argc, argv);

   name = DStrAlloc();

   /* Step 1: Read existing files: problems, clausepatterns, signature
    */
   proof_examples = ExampleSetAlloc();
   in = CreateScanner(StreamTypeFile,
                      KBFileName(name, kb_name, "problems"),
                      true, NULL, true);
   ExampleSetParse(in, proof_examples);
   DestroyScanner(in);

   typebank = TypeBankAlloc();
   reserved_symbols = SigAlloc(typebank);

   in = CreateScanner(StreamTypeFile,
                      KBFileName(name, kb_name, "signature"),
                      true, NULL, true);
   SigParse(in, reserved_symbols, true);
   DestroyScanner(in);

   in = CreateScanner(StreamTypeFile,
                      KBFileName(name, kb_name, "clausepatterns"),
                      true, NULL, true);
   annoterms = TBAlloc(reserved_symbols);
   clause_examples = AnnoSetParse(in, annoterms, KB_ANNOTATION_NO);
   DestroyScanner(in);

   VERBOUT("Old knowledge base files parsed successfully\n");

   if(state->argc ==  0)
   {
      CLStateInsertArg(state, "-");
   }

   /* Loop over all new examples and integrate them */

   for(i=0; state->argv[i]; i++)
   {

      /* Step 2: Determine name and check validity */

      if(!ex_name && state->argv[i] && (strcmp(state->argv[i], "-")!= 0))
      {
         ex_name = FileFindBaseName(state->argv[i]);
      }

      if(!ex_name)
      {
         sprintf(defaultname, "__problem__%ld",
                 proof_examples->count+1);
         ex_name = defaultname;
      }

      if(ExampleSetFindName(proof_examples, ex_name))
      {
         DStr_p error = DStrAlloc();

         DStrAppendStr(error, "Example name '");
         DStrAppendStr(error, ex_name);
         DStrAppendStr(error, "' already in use");
         Error(DStrView(error), USAGE_ERROR);

         DStrFree(error);
      }

      VERBOUTARG("New example will use name ", ex_name);

      /* Step 3: Copy input file into files-directory */

      store_file = DStrAlloc();
      DStrAppendStr(store_file, KBFileName(name, kb_name, "FILES/"));
      DStrAppendStr(store_file, ex_name);

      CopyFile(DStrView(store_file), state->argv[i]);

      /* Step 4: Integrate new examples into existing structures */

      in = CreateScanner(StreamTypeFile, DStrView(store_file), true, NULL, true);

      KBParseExampleFile(in, ex_name, proof_examples, clause_examples,
                         reserved_symbols);
      DestroyScanner(in);
      DStrFree(store_file);
      ex_name = NULL;
   }

   /* Step 5: Write everything back: problems, clausepatterns */

   out = OutOpen(KBFileName(name, kb_name, "clausepatterns"));
   AnnoSetPrint(out, clause_examples);
   OutClose(out);

   out = OutOpen(KBFileName(name, kb_name, "problems"));
   ExampleSetPrint(out, proof_examples);
   OutClose(out);

   /* Finally clean up */

   DStrFree(name);
   AnnoSetFree(clause_examples);
   annoterms->sig = NULL;
   TBFree(annoterms);
   SigFree(reserved_symbols);
   TypeBankFree(typebank);
   ExampleSetFree(proof_examples);
   CLStateFree(state);

   ExitIO();
#ifdef CLB_MEMORY_DEBUG
   MemFlushFreeList();
   MemDebugPrintStats(stdout);
#endif

   return 0;
}


/*-----------------------------------------------------------------------
//
// Function: process_options()
//
//   Read and process the command line option, return (the pointer to)
//   a CLState object containing the remaining arguments.
//
// Global Variables:
//
// Side Effects    : Sets variables, may terminate with program
//                   description if option -h or --help was present
//
/----------------------------------------------------------------------*/

CLState_p process_options(int argc, char* argv[])
{
   Opt_p handle;
   CLState_p state;
   char*  arg;

   state = CLStateAlloc(argc,argv);

   while((handle = CLStateGetOpt(state, &arg, opts)))
   {
      switch(handle->option_code)
      {
      case OPT_VERBOSE:
            Verbose = CLStateGetIntArg(handle, arg);
            break;
      case OPT_HELP:
            print_help(stdout);
            exit(NO_ERROR);
      case OPT_VERSION:
            printf(NAME " " VERSION "\n");
            exit(NO_ERROR);
      case OPT_KB:
            kb_name = arg;
            break;
      case OPT_NAME:
            ex_name = arg;
            break;
      default:
            assert(false);
            break;
      }
   }
   return state;
}

void print_help(FILE* out)
{
   fprintf(out, "\n\
\n"
           NAME " " VERSION "\n\
\n\
Usage: ekb_insert [options] [names]\n\
\n\
Insert example files into an E knowledge base. Each non-option argument\n\
is considered as one individual example file. For most applications\n\
this is obsolete, use ekb_ginsert instead.\n\n");
   PrintOptions(stdout, opts, "Options\n\n");
   fprintf(out, "\n\n" E_FOOTER);
}


/*---------------------------------------------------------------------*/
/*                        End of File                                  */
/*---------------------------------------------------------------------*/
